### PR TITLE
Prometheus Remote Read ETL tool

### PR DESCRIPTION
**Pull request by joshmeranda**
OEP: https://github.com/rancher/opni/pull/970

### CODE REVIEW

**Review by kralicky — CHANGES_REQUESTED** [review]
Looks good so far. Couple of things to do before we can merge:
- There are a lot of todos, you should attempt to resolve as many as possible (all the ones that aren't blocked by other things)
- There are a couple of style related comments, do a once over through all the new code and make sure code style stays consistent.

**Comment by joshmeranda** [comment]
Sounds good, adding a few more tests I thought of while  writing docs, so I'll add those, and start addressing your comments. Thanks!

File: plugins/metrics/pkg/agent/http.go
Comment (alexandreLamarre) [inline plugins/metrics/pkg/agent/http.go:137]: Do these commented out functions serve a purpose?

Comment (joshmeranda) [reply plugins/metrics/pkg/agent/http.go:137]: nope, I definitely just commented those out when I was still tinkering and forgot about them. I'll delete them

Comment (joshmeranda) [reply plugins/metrics/pkg/agent/http.go:137]: nope, I definitely just commented those out when I was still tinkering and forgot about them. I'll delete them now

File: plugins/metrics/pkg/agent/node.go
Comment (kralicky) [inline plugins/metrics/pkg/agent/node.go:206]: You should add the `Discover` method to the driver interface, feel free to refactor as necessary (it'll probably look weird but in this case it's okay for the driver interface to be a collection of unrelated methods). Maybe also rename to `Discover{RemoteReadTargets|Targets|Prometheuses|etc}`.

Adding the method will also require an implementation in the testenv driver, which will definitely be useful for integration tests.

File: plugins/metrics/pkg/agent/node.go
Comment (kralicky) [inline plugins/metrics/pkg/agent/node.go:233]: Is it possible to get here? Looks like doSync is only called if nodeClient is not nil in the first place.

File: plugins/metrics/pkg/apis/remoteread/client.go
Comment (kralicky) [inline plugins/metrics/pkg/apis/remoteread/client.go:36]: Suggest renaming to convey that this is specifically for the gateway client

File: plugins/metrics/pkg/backend/metrics.go
Comment (kralicky) [inline plugins/metrics/pkg/backend/metrics.go:447]: Generally, errors returned from grpc handler functions should be of the form `status.Error(codes.XXX, ...)`, otherwise the grpc status code gets set to Unknown which makes it harder to handle errors on the client side.
```suggestion
		return nil, status.Errorf(codes.NotFound, "target not found: %s", id)
```

File: plugins/metrics/pkg/cortex/api.go
Comment (kralicky) [inline plugins/metrics/pkg/cortex/api.go:28]: Unused?

**Review by kralicky — CHANGES_REQUESTED** [review]
Might also need to rebase

File: pkg/opni/util/render_cortex.go
Comment (joshmeranda) [reply pkg/opni/util/render_cortex.go:213]: you're right, when I added this I was getting weird nil entries in the received  `TargetList`

Comment (kralicky) [reply pkg/opni/util/render_cortex.go:213]: If you are still getting that issue, it's likely a bug on the server side

File: plugins/metrics/pkg/cortex/api.go
Comment (joshmeranda) [reply plugins/metrics/pkg/cortex/api.go:28]: yep not used, I'm gonna do another once over to catch anything that I forget about when I was doing the initial integration and didn't know where things went

File: plugins/metrics/pkg/agent/node.go
Comment (joshmeranda) [reply plugins/metrics/pkg/agent/node.go:233]: I think the only change I made here was accidentally chaning the `client` -> `promClient`.

Regardless I think in [MetricsNode.SetNodeClient](https://github.com/rancher/opni/blob/ff5c37f4522a74ad086bf86208f53fd876d78238/plugins/metrics/pkg/agent/node.go#L95-L102) its possible to pass a `nil` client. The only place its called is with a non-nil client, but its *possible*. It should be safe to remove the check, but I'd rather leave it in or move the check into `SetNodeClient`.

File: plugins/metrics/pkg/agent/node.go
Comment (kralicky) [inline plugins/metrics/pkg/agent/node.go:211]: The driver here still needs to be dependency-injected. I suggest handling it somewhere in [here](https://github.com/rancher/opni/blob/main/plugins/metrics/pkg/agent/plugin.go#L34-L54) maybe by passing the driver to NewMetricsNode.

**Review by kralicky — APPROVED** [review]

**Review by kralicky — DISMISSED** [review]

**Pull request merged**
